README
Full documentation for VeryNice is in the "verynice.html" file, or
on the web at http://www.tam.cornell.edu/~sdh4/verynice/



CHANGELOG
VeryNice 0.8
------------
A working SysV style init script is now provided. 
-d [pidfile] option for automatic process-id file generation added. 
"hungryexe" option added
VeryNice now only looks up per-user configuration information for users
who are actually logged in, not all users with accounts (dramatic scalability
improvement). 
cpu count detection is now properly OS-dependent. Warning message about 
lack of cpu count support added for Solaris
We got SIGKILL and SIGTERM backwards the first time around, so processes
weren't getting a cahnce to clean themselves up before being killed.


VeryNice 0.5 
------------
First release

config.h
#ifndef CONFIG_H
#define CONFIG_H

void init_config(void);
void config(void);
typedef union {
  long alignment;
  char ag_vt_2[sizeof(int)];
  char ag_vt_4[sizeof(char)];
  char ag_vt_5[sizeof(double)];
} config_vs_type;

typedef enum {
  config_ConfigFile_token = 1, config_lines_token, config_EOF_token,
  config_ws_token, config_line_token = 7, config_EOL_token,
  config_commentchars_token = 10, config_unquotedstring_token = 14,
  config_quotedstring_token = 16, config_signed_integer_token = 22,
  config_unsigned_real_token = 27, config_unsigned_integer_token = 30,
  config_unquotedstringchar_token = 32, config_quotedstringchars_token = 34,
  config_RegularChar_token, config_EscapeSequence_token,
  config_DIGIT_token = 42, config_mantissa_token, config_EXPSYM_token,
  config_integer_part_token, config_fraction_part_token = 47
} config_token_type;

typedef struct {
  config_token_type token_number, reduction_token, error_frame_token;
  int input_code;
  int input_value;
  int line, column;
  int ssx, sn, error_frame_ssx;
  int drt, dssx, dsn;
  int ss[32];
  config_vs_type vs[32];
  int bts[32], btsx;
  unsigned char * pointer;
  unsigned char * la_ptr;
  int lab[24], rx, fx;
  const unsigned char *key_sp;
  int save_index, key_state;
  char *error_message;
  char read_flag;
  char exit_flag;
} config_pcb_type;


#ifndef PRULE_CONTEXT
#define PRULE_CONTEXT(pcb)  (&((pcb).cs[(pcb).ssx]))
#define PERROR_CONTEXT(pcb) ((pcb).cs[(pcb).error_frame_ssx])
#define PCONTEXT(pcb)       ((pcb).cs[(pcb).ssx])
#endif


#ifndef AG_RUNNING_CODE_CODE
/* PCB.exit_flag values */
#define AG_RUNNING_CODE         0
#define AG_SUCCESS_CODE         1
#define AG_SYNTAX_ERROR_CODE    2
#define AG_REDUCTION_ERROR_CODE 3
#define AG_STACK_ERROR_CODE     4
#define AG_SEMANTIC_ERROR_CODE  5
#endif

extern config_pcb_type config_pcb;
#endif



linklist.c
/*
     VeryNice -- a dynamic process re-nicer
     Copyright (C) 1992-1999 Stephen D. Holland
 
     This library is free software; you can redistribute it and/or
     modify it under the terms of the GNU Lesser General Public
     License as published by the Free Software Foundation; 
     version 2.1 of the License.

     This library is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Lesser General Public License for more details.

     You should have received a copy of the GNU Lesser General Public
     License along with this library; if not, write to the Free Software
     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
*/

#include <string.h>
#include "linklist.h"

/* linklist.c -- This API is derived from the link list routines 
   that were included as part of the original Amiga operating system.

   The basic idea is to provide a doubly linked-list system with a 
   minimum of special cases required in the handling code. This is
   achieved by having two permanent members ("nodes") of the list, 
   called the Head and the Tail, that are never removed and have no
   actual data associated with them. Because these members are permanent,
   there are no special cases involved with removing or adding the first
   or last nodes of the list. 

   The head and tail nodes are contained within the "List" structure, and
   actually overlap! The "Predecessor" field of the head node is the 
   same as the "Successor" field of the tail node (both are always NULL). 

   Lists must be initialized with NewList() before they can be used. */


void Insert(struct List *ThisList,struct Node *AddNode,struct Node *Pred)
{
	if (AddNode==NULL) return;
	if (Pred==NULL) {
		Pred=(struct Node *)&ThisList->lh_Head;
	}
	AddNode->ln_Pred=Pred;
	AddNode->ln_Succ=Pred->ln_Succ;
	Pred->ln_Succ=AddNode;
	AddNode->ln_Succ->ln_Pred=AddNode;
}

void Remove(struct Node *RemNode)
{
	RemNode->ln_Pred->ln_Succ=RemNode->ln_Succ;
	RemNode->ln_Succ->ln_Pred=RemNode->ln_Pred;
	RemNode->ln_Pred=NULL;
	RemNode->ln_Succ=NULL;
}

void AddHead(struct List *ThisList,struct Node *AddNode)
{
	Insert(ThisList,AddNode,NULL);
}

struct Node *RemHead(struct List *ThisList)
{
	struct Node *Head;
	Head=ThisList->lh_Head;
	if (Head->ln_Succ==NULL) {
		return NULL;
	}
	Remove(Head);
	return Head;
}

void AddTail(struct List *ThisList,struct Node *AddNode)
{
	Insert(ThisList,AddNode,ThisList->lh_TailPred);
}

struct Node *RemTail(struct List *ThisList)
{
	struct Node *Tail;
	Tail=ThisList->lh_TailPred;
	if (Tail->ln_Pred==NULL) {
		return NULL;
	}
	Remove(Tail);
	return Tail;
}

void Enqueue(struct List *ThisList,struct Node *AddNode)
{
	struct Node *CurrentSucc;
	for (CurrentSucc=ThisList->lh_Head;
		(CurrentSucc->ln_Succ!=NULL)&&(CurrentSucc->ln_Pri >= AddNode->ln_Pri);
		CurrentSucc=CurrentSucc->ln_Succ) ;
	Insert(ThisList,AddNode,CurrentSucc->ln_Pred);
}

struct Node *FindName(struct List *ThisList,char *Name)
{
	struct Node *CurrentNode;
	for (CurrentNode=ThisList->lh_Head;
		CurrentNode->ln_Succ->ln_Succ != NULL;
		CurrentNode=CurrentNode->ln_Succ) {
			if (strcmp(CurrentNode->ln_Name,Name)==0) {
				return CurrentNode;
			}
	}
	return NULL;
}

void NewList(struct List *List)
/* set up a list for use */
{
	if (!List) return;
	List->lh_Head=(struct Node *)&List->lh_Tail;
	List->lh_Tail=NULL;
	List->lh_TailPred=(struct Node *)&List->lh_Head;
	List->lh_Type=0;
}



linklist.h
/*
     VeryNice -- a dynamic process re-nicer
     Copyright (C) 1992-1999 Stephen D. Holland
 
     This library is free software; you can redistribute it and/or
     modify it under the terms of the GNU Lesser General Public
     License as published by the Free Software Foundation; 
     version 2.1 of the License.

     This library is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Lesser General Public License for more details.

     You should have received a copy of the GNU Lesser General Public
     License along with this library; if not, write to the Free Software
     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA

*/
/* See linklist.c for documentation */

#ifndef NULL
#define NULL (void *)0L
#endif

struct Node {
	struct Node *ln_Succ,*ln_Pred;
	unsigned char ln_Type;
	char ln_Pri,*ln_Name;
};

struct List {
	struct Node *lh_Head,*lh_Tail,*lh_TailPred;
	unsigned char lh_Type,l_pad;
};

struct MinNode {
	struct MinNode *mln_Succ,*mln_Pred;
};

struct MinList {
	struct MinNode *mlh_Head, *mlh_Tail,*mlh_TailPred;
};

void Insert(struct List *ThisList,struct Node *AddNode,struct Node *Pred);
void Remove(struct Node *RemNode);
void AddHead(struct List *ThisList,struct Node *AddNode);
struct Node *RemHead(struct List *ThisList);
void AddTail(struct List *ThisList,struct Node *AddNode);
struct Node *RemTail(struct List *ThisList);
void Enqueue(struct List *ThisList,struct Node *AddNode);
struct Node *FindName(struct List *ThisList,char *Name);
void NewList(struct List *List);


verynice.h
/*
     VeryNice -- a dynamic process re-nicer
     Copyright (C) 2000 Stephen D. Holland
 
     This program is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation; version 2 of the License.
 
     This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU General Public License for more details.
 
     You should have received a copy of the GNU General Public License
     along with this program; if not, write to the Free Software
     Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


struct procent {
  struct Node Node;
  pid_t pid;
  int nicelevel; /* actual nice level of process, including amount reniced by
		    verynice */
  int reniced; /* amount reniced by verynice(tm) */
  pid_t parentpid;
  double cpuusage; /* total cpuusage up to now of process, in seconds */

  double current_cpuusage; /* current delta of cpuusage */
  
  double badkarma; /* accumulated bad karma of process */
  uid_t uid;
  int immuneflag; /* pid and children of same uid are immune */
  int goodflag; /* pid and children are 'good' -- should automatically 
		   be reniced negatively, regardless of CPU usage */
  int badflag; /* pid and children are 'bad'  -- should be niced automatically,
		  regardless of CPU usage */
  int potentialrunaway; /* very heavy CPU usage will cause this to be 
			   set to RUNAWAY level */
  int hungryflag; /* Always assume 100% cpu usage (not inherited) */ 
  char *exename; /* name of executable, separate malloc */
  int reallynew; /* set if this is really a new process or not */
  int deadflag; /* waiting to be expunged */
};

extern struct List proclist;
extern struct List newproclist;

struct immuneuid {
  struct Node Node;
  uid_t uid; /* these uid are immune */

};

extern struct List iuidlist;

struct immuneexe {
  struct Node Node;
  char *exename; /* name or substring of executable, separate malloc */
  int alluid; /* 1 if this applys to all uids, 0 if it applys only to uid,
		 below */
  uid_t uid;
};

extern struct List iexelist;

struct badexe {
  struct Node Node;
  char *exename; /* name or substring of executable, separate malloc */  
  int alluid; /* 1 if this applys to all uids, 0 if it applys only to uid,
		 below */
  uid_t uid;
};
extern struct List bexelist;

struct goodexe { /* for multimedia apps, which should be negatively niced */
  struct Node Node;
  char *exename; /* name or substring of executable, separate malloc */  
  int alluid; /* 1 if this applys to all uids, 0 if it applys only to uid,
		 below */
  uid_t uid;
};
extern struct List gexelist;


struct runawayexe { /* for apps that tend to get into runaway CPU usage
		       (e.g. netscape) */
  struct Node Node;
  char *exename; /* name or substring of executable, separate malloc */  
  int alluid; /* 1 if this applys to all uids, 0 if it applys only to uid,
		 below */
  uid_t uid;
};
extern struct List runawaylist;

struct hungryexe { 
  /* for apps that are CPU hungry whether or not they appear to be
     appear to be (e.g. make) */
  struct Node Node;
  char *exename;
  int alluid; /* 1 if this applys to all uids, 0 if it applys only to uid,
		 below */
  uid_t uid;
  
};

extern struct List hungrylist;

struct knownuid { /* track the UID's for which we've attempted to read
		     config information */
  struct Node Node;
  uid_t uid;
};
extern struct List uidlist;


#define GLOBALCFGFILE "verynice.conf"
#define USERCFGFILE ".verynicerc"

extern int normal;
extern int notnice;
extern int batchjob;
extern int runaway;
extern int killproc;

extern int periodicity;
extern int rereadcfgperiodicity;

extern double badkarmarate;
extern double karmarestorationrate;


/* WARNING: side effects:
       - Assumes syslog(3) is activated. Caller is responsible for opening/closing
       - May open password database if not already open, and adjust location
       - This routine closes password database
  */
void ReadCfgFiles(char *prefix);


struct knownuid *finduid(uid_t uid);


/* caller responsible for calling endpwent() after this */
/* this reads in the config file, and adds an appropriate entry
   to the known uid list 
   (returns pointer to entry in known uid list) */
struct knownuid *ReadCfgFile(uid_t uid);



stringstack.h
/*
     VeryNice -- a dynamic process re-nicer
     Copyright (C) 1992-1999 Stephen D. Holland

     This library is free software; you can redistribute it and/or
     modify it under the terms of the GNU Lesser General Public
     License as published by the Free Software Foundation; 
     version 2.1 of the License.

     This library is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Lesser General Public License for more details.

     You should have received a copy of the GNU Lesser General Public
     License along with this library; if not, write to the Free Software
     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA

*/


struct Stack {
	struct MinList StringList;
	char *PopStringBuf;
	long PopStringBufSize; /* Allocated size */
};

struct StkNode {
	struct MinNode Node;
	char *String; /* String actually stored in memory after StkNode */
	int AllocLen; /* Allocated length, including StkNode and string */
};

#define SETSTRING(StkNode) {((struct StkNode *)(StkNode))->String=((char *)(StkNode))+sizeof(struct StkNode);}
#define STRINCREMENT 150 /* incremental string size -- between realloc()s */

/* Push a string into a new stack entry */
/* returns NULL in case of error or original string pointer */
char * PushString(struct Stack *Stk,char *Str);

/*
	remove and copy into static storage the top string on the stack.
	NOTE: pointer returned will become invalid the next time PopString()
	is called with the same stack pointer
	returns NULL in case of error.
*/
char *PopString(struct Stack *Stk); 

/* Add a character to the end of the string at the top of the stack */
/* Returns character, or -1 for error */
int AddChar(struct Stack *Stk,int Chr);

/* Add a string to the end of the string at the top of the stack */
/* Returns original string pointer, or NULL for error */
char *AddStr(struct Stack *Stk,char *Str);

/* Get pointer to the string on the top of the stack. This pointer
becomes invalid if AddChar or AddStr is called on that member of the stack.
IT REMAINS VALID IF InitString() IS CALLED AND AddChar/AddStr is used on
another member of the stack */
/* returns NULL in case of error */
char *GetString(struct Stack *Stk);

/* Initialize a string on the top of the stack */
/* returns NULL in case of error */
char *InitString(struct Stack *Stk) ;

/* Insert Str at the beginning of the string on the top of the stack */
void InsertString(struct Stack *Stk,char *Str);

/* Initialize a stack */
/* returns NULL in case of error */
struct Stack *InitStack(void);

/* close down a stack */
void UnitStack(struct Stack *Stk);

/* Count the number of entries on a stack */
int CountStack(struct Stack *Stk);


stringstack.c

/*
     VeryNice -- a dynamic process re-nicer
     Copyright (C) 1992-1999 Stephen D. Holland
 
     This library is free software; you can redistribute it and/or
     modify it under the terms of the GNU Lesser General Public
     License as published by the Free Software Foundation; 
     version 2.1 of the License.

     This library is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Lesser General Public License for more details.

     You should have received a copy of the GNU Lesser General Public
     License along with this library; if not, write to the Free Software
     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA

*/


#define STDCALLOC
#ifdef STDCALLOC
#define MALLOCFUNC(size) malloc(size)
#define CALLOCFUNC(size) calloc(1,(size))
#define FREEFUNC(ptr,size) free(ptr)
#else
#ifdef AMIGAEXECALLOC
#define MALLOCFUNC(size) AllocMem((size),MEMF_PUBLIC)
#define CALLOCFUNC(size) AllocMem((size),MEMF_PUBLIC|MEMF_CLEAR)
#define FREEFUNC(ptr,size) FreeMem((ptr),(size))
#else
#error StringStack needs to know which allocation functions to use!
#endif
#endif


#include <string.h>
#include <stdlib.h>

#include "linklist.h"
#include "stringstack.h"

#if (STRINCREMENT < 1)
	#error Zero string increment!
#endif

static struct StkNode *GetTopStkNode(struct Stack *Stk)
{
	if (!Stk) return NULL;
	
	if (Stk->StringList.mlh_TailPred == (struct MinNode *)&Stk->StringList.mlh_Head)
		return NULL; /* Empty list */
	return (struct StkNode *)Stk->StringList.mlh_TailPred;
}

/* Push a string into a new stack entry */
char *PushString(struct Stack *Stk,char *Str)
{
	if (!Stk || !Str) return NULL;
	
	if (!InitString(Stk)) return NULL;
	if (!AddStr(Stk,Str)) return NULL;
	return Str;
}

/*
	remove and copy into static storage the top string on the stack.
	NOTE: pointer returned will become invalid the next time PopString()
	is called with the same stack pointer 
*/
char *PopString(struct Stack *Stk)
{
	struct StkNode *StkNode;
	int Len;
	
	if (!Stk) return NULL;
	
	StkNode=(struct StkNode *)RemTail((struct List *)&Stk->StringList);
	if (!StkNode) return NULL;
	
	Len=strlen(StkNode->String);
	if (Stk->PopStringBuf && Stk->PopStringBufSize >= (Len+1)) {
		strcpy(Stk->PopStringBuf,StkNode->String);
		FREEFUNC(StkNode,StkNode->AllocLen);
		return Stk->PopStringBuf;
	} else {
		if (Stk->PopStringBuf) {
			FREEFUNC(Stk->PopStringBuf,Stk->PopStringBufSize);
		}
		Stk->PopStringBuf=(char *)MALLOCFUNC(Stk->PopStringBufSize=(Len+1));
		if (!Stk->PopStringBuf) {FREEFUNC(StkNode,StkNode->AllocLen);return NULL;}
		strcpy(Stk->PopStringBuf,StkNode->String);
		FREEFUNC(StkNode,StkNode->AllocLen);
		return Stk->PopStringBuf;
	}
	
}

/* Add a character to the end of the string at the top of the stack */
/* Returns character, or -1 for error */
int AddChar(struct Stack *Stk,int Chr)
{
	struct StkNode *StkNode,*OldNode;
	struct MinNode *Pred;
	int Len,ExtraSpace,SpaceAlloc;
	
	if (!Stk) return -1;
	
	StkNode=GetTopStkNode(Stk);
	if (!StkNode) return -1;
	
	Len=strlen(StkNode->String);
	
	ExtraSpace=StkNode->AllocLen-sizeof(struct StkNode)-Len-1-1;
	if (ExtraSpace < 0) {
		SpaceAlloc=((((-ExtraSpace*STRINCREMENT)/STRINCREMENT)+STRINCREMENT)) + StkNode->AllocLen;
		Pred=StkNode->Node.mln_Pred;
		Remove((struct Node *)StkNode);
		StkNode=(struct StkNode *)realloc(OldNode=StkNode,SpaceAlloc);
		if (!StkNode) { /* if error, put it back on list and return error */
			Insert((struct List *)&Stk->StringList,(struct Node *)OldNode,(struct Node *)Pred);
			return -1;
		}
		Insert((struct List *)&Stk->StringList,(struct Node *)StkNode,(struct Node *)Pred);
    StkNode->AllocLen=SpaceAlloc;
    SETSTRING(StkNode);
	}
	StkNode->String[strlen(StkNode->String)+1]=0; /* Null terminator */
	StkNode->String[strlen(StkNode->String)]=Chr;
	return Chr;
	
}

/* Add a string to the end of the string at the top of the stack */
/* Returns original string pointer, or NULL for error */
char *AddStr(struct Stack *Stk,char *Str)
{
	struct StkNode *StkNode,*OldNode;
	struct MinNode *Pred;
	int Len,ExtraSpace,SpaceAlloc;
	
	if (!Stk || !Str) return NULL;
	
	StkNode=GetTopStkNode(Stk);
	if (!StkNode) return NULL;
	
	Len=strlen(StkNode->String);
	
	ExtraSpace=StkNode->AllocLen-sizeof(struct StkNode)-Len-1-strlen(Str);
	if (ExtraSpace < 0) {
		SpaceAlloc=((((-ExtraSpace*STRINCREMENT)/STRINCREMENT)+STRINCREMENT)) + StkNode->AllocLen;
		Pred=StkNode->Node.mln_Pred;
		Remove((struct Node *)StkNode);
		StkNode=(struct StkNode *)realloc(OldNode=StkNode,SpaceAlloc);
		if (!StkNode) { /* if error, put it back on list and return error */
			Insert((struct List *)&Stk->StringList,(struct Node *)OldNode,(struct Node *)Pred);
			return NULL;
		}
		Insert((struct List *)&Stk->StringList,(struct Node *)StkNode,(struct Node *)Pred);
    StkNode->AllocLen=SpaceAlloc;
    SETSTRING(StkNode);
	}
	strcat(StkNode->String,Str);
	return Str;
}

/* Get pointer to the string on the top of the stack. This pointer
becomes invalid if AddChar or AddStr is called on the stack */
char *GetString(struct Stack *Stk)
{
	struct StkNode *StkNode;
	if (!Stk) return NULL;
	
	StkNode=GetTopStkNode(Stk);
	if (!StkNode) return NULL;
	return StkNode->String;
}

/* Initialize a string on the top of the stack */
char *InitString(struct Stack *Stk)
{
	struct StkNode *StkNode;
	if (!Stk) return NULL;
	
	StkNode=(struct StkNode *)CALLOCFUNC(sizeof(struct StkNode)+STRINCREMENT);
	if (!StkNode) return NULL;
	StkNode->AllocLen=sizeof(struct StkNode)+STRINCREMENT;
	SETSTRING(StkNode);
	StkNode->String[0]='\0';
	AddTail((struct List *)&Stk->StringList,(struct Node *)StkNode);
	return StkNode->String;
}

/* close down a stack */
void UnitStack(struct Stack *Stk)
{
	if (!Stk) return;
	while (PopString(Stk)) ; /* pop all strings off the stack */
	if (Stk->PopStringBuf) FREEFUNC(Stk->PopStringBuf,Stk->PopStringBufSize);
	FREEFUNC(Stk,sizeof(struct Stack));
}

/* Initialize a stack */
struct Stack *InitStack(void)
{
	struct Stack *Stk;
	
	Stk=(struct Stack *)CALLOCFUNC(sizeof(struct Stack));
	if (!Stk) return NULL;
	Stk->PopStringBuf=NULL;
	Stk->PopStringBufSize=0;
	NewList((struct List *)&Stk->StringList);
	return Stk;
}

int CountStack(struct Stack *Stk)
     /* Count the number of entries on a stack */
{
  struct MinNode *Node;
  int Cnt=0;

  for (Node=Stk->StringList.mlh_Head;Node->mln_Succ;Node=Node->mln_Succ)
    Cnt++;
  return Cnt;
}
